Mathlib/AlgebraicGeometry/SpreadingOut.lean

```diff
@@ -0,0 +1,293 @@
+    rw [X.presheaf.germ_res_apply, ‹φRA ≫ φ = _›]
+    rfl
+
+/--
+Given `S`-schemes `X Y` and points `x : X` `y : Y` over `s : S`.
+Suppose we have the following diagram of `S`-schemes
+```
+Spec 𝒪_{X, x} ⟶ X
+    |
+  Spec(φ)
+    ↓
+Spec 𝒪_{Y, y} ⟶ Y
+```
+Then the map `Spec(φ)` spreads out to an `S`-morphism on an open subscheme `U ⊆ X`,
+```
+Spec 𝒪_{X, x} ⟶ U ⊆ X
+    |             |
+  Spec(φ)         |
+    ↓             ↓
+Spec 𝒪_{Y, y} ⟶ Y
+```
+provided that `Y` is locally of finite type over `S` and
+`X` is "germ-injective" (e.g. when it's integral or locally noetherian).
+
+TODO: The condition on `X` is unnecessary when `Y` is locally of finite presentation.
+-/
+@[stacks 0BX6]
+lemma spread_out_of_germInjective [LocallyOfFiniteType sY] [X.GermInjective] {x : X} {y : Y}
+    (e : sX.base x = sY.base y) (φ : Y.presheaf.stalk y ⟶ X.presheaf.stalk x)
+    (h : sY.stalkMap y ≫ φ =
+      S.presheaf.stalkSpecializes (Inseparable.of_eq e).specializes ≫ sX.stalkMap x) :
+    ∃ (U : X.Opens) (hxU : x ∈ U) (f : U.toScheme ⟶ Y),
+      Spec.map φ ≫ Y.fromSpecStalk y = U.fromSpecStalkOfMem x hxU ≫ f ∧
+      f ≫ sY = U.ι ≫ sX := by
+  obtain ⟨_, ⟨U, hU, rfl⟩, hxU, -⟩ :=
+    (isBasis_affine_open S).exists_subset_of_mem_open (Set.mem_univ (sX.base x)) isOpen_univ
+  have hyU : sY.base y ∈ U := e ▸ hxU
+  obtain ⟨_, ⟨V : Y.Opens, hV, rfl⟩, hyV, iVU⟩ :=
+    (isBasis_affine_open Y).exists_subset_of_mem_open hyU (sY ⁻¹ᵁ U).2
+  have : sY.appLE U V iVU ≫ Y.presheaf.germ V y hyV ≫ φ =
+      sX.app U ≫ X.presheaf.germ (sX ⁻¹ᵁ U) x hxU := by
+    rw [Scheme.Hom.appLE, Category.assoc, Y.presheaf.germ_res_assoc,
+      ← Scheme.stalkMap_germ_assoc, h]
+    simp
+  obtain ⟨W, hxW, φ', i, hW, h₁, h₂⟩ :=
+    exists_lift_of_germInjective (R := Γ(S, U)) (A := Γ(Y, V)) (U := sX ⁻¹ᵁ U) (x := x) hxU
+    (Y.presheaf.germ _ y hyV ≫ φ) (sY.appLE U V iVU) (sX.app U)
+    (LocallyOfFiniteType.finiteType_of_affine_subset ⟨_, hU⟩ ⟨_, hV⟩ _) this
+  refine ⟨W, hxW, W.toSpecΓ ≫ Spec.map φ' ≫ hV.fromSpec, ?_, ?_⟩
+  · rw [W.fromSpecStalkOfMem_toSpecΓ_assoc x hxW, ← Spec.map_comp_assoc, ← h₁,
+      Spec.map_comp, Category.assoc, ← IsAffineOpen.fromSpecStalk,
+      IsAffineOpen.fromSpecStalk_eq_fromSpecStalk]
+  · simp only [Category.assoc, IsAffineOpen.isoSpec_inv_ι_assoc]
+    rw [← IsAffineOpen.Spec_map_appLE_fromSpec sY hU hV iVU, ← Spec.map_comp_assoc, ← h₂,
+      ← Scheme.Hom.appLE, ← hW.isoSpec_hom, IsAffineOpen.Spec_map_appLE_fromSpec sX hU hW i,
+      ← Iso.eq_inv_comp, IsAffineOpen.isoSpec_inv_ι_assoc]
+
+/--
+Let `x : X` and `f g : X ⟶ Y` be two morphisms such that `f x = g x`.
+If `f` and `g` agree on the stalk of `x`, then they agree on an open neighborhood of `x`,
+provided `X` is "germ-injective" (e.g. when it's integral or locally noetherian).
+
+TODO: The condition on `X` is unnecessary when `Y` is locally of finite type.
+-/
+@[stacks 0BX6]
+lemma spread_out_unique_of_germInjective [X.GermInjective] {x : X}
```

Comment:
It seems this is a more basic lemma, and that it uses only `X.IsGermInjectiveAt x`, which may also be the take of of the other lemmas above.

Reply:
Apart from the `X.IsGermInjectiveAt x` part of the comment which I agree is a great suggestion, I'm not sure what you meant by "a more basic lemma". Are you suggesting that there is a more basic generalization or that it should be moved forwards in the file or something else?

Reply:
Yes, the lemma could be moved upwards as it seems more basic.